src/google/protobuf/descriptor.cc

```diff
@@ -5487,6 +5497,112 @@ void DescriptorBuilder::BuildMessage(const DescriptorProto& proto,
   }
 }
 
+void DescriptorBuilder::CheckFieldJsonNameUniqueness(
+    const DescriptorProto& proto, const Descriptor* result) {
+  FileDescriptor::Syntax syntax = result->file()->syntax();
+  std::string message_name = result->full_name();
+  // two passes: one looking only at default JSON names, and one that considers
+  // custom JSON names
+  CheckFieldJsonNameUniqueness(message_name, proto, syntax, false);
+  CheckFieldJsonNameUniqueness(message_name, proto, syntax, true);
+}
+
+namespace {
+// Helpers for function below
+
+struct JsonNameDetails {
+  const FieldDescriptorProto* field;
+  std::string orig_name;
+  bool is_custom;
+};
+
+JsonNameDetails GetJsonNameDetails(const FieldDescriptorProto* field, bool use_custom) {
+  std::string default_json_name = ToJsonName(field->name());
+  if (use_custom && field->has_json_name() &&
+      field->json_name() != default_json_name) {
+    return {field, field->json_name(), true};
+  }
+  return {field, default_json_name, false};
+}
+
+bool JsonNameLooksLikeExtension(std::string name) {
+  return !name.empty() && name.front() == '[' && name.back() == ']';
+}
+
+
+} // namespace
+
+void DescriptorBuilder::CheckFieldJsonNameUniqueness(
+    const std::string& message_name, const DescriptorProto& message,
+    FileDescriptor::Syntax syntax, bool use_custom_names) {
+
+  absl::flat_hash_map<std::string, JsonNameDetails> name_to_field;
+  for (const FieldDescriptorProto& field : message.field()) {
+    JsonNameDetails details = GetJsonNameDetails(&field, use_custom_names);
+    if (details.is_custom && JsonNameLooksLikeExtension(details.orig_name)) {
+      std::string error_message =
+          absl::StrFormat("The custom JSON name of field \"%s\" (\"%s\") is invalid: "
+                          "JSON names may not start with '[' and end with ']'.",
+                          field.name(), details.orig_name);
+      AddError(message_name, field, DescriptorPool::ErrorCollector::NAME,
+               error_message);
+      continue;
+    }
+
+    std::string lowercase_name = absl::AsciiStrToLower(details.orig_name);
+    auto it_inserted = name_to_field.try_emplace(lowercase_name, details);
+    if (it_inserted.second) {
+      continue;
+    }
+    JsonNameDetails& match = it_inserted.first->second;
+    if (use_custom_names && !details.is_custom && !match.is_custom) {
+      // if this pass is considering custom JSON names, but neither of the
+      // names involved in the conflict are custom, don't bother with a
+      // message. That will have been reported from other pass (non-custom
+      // JSON names).
+      continue;
+    }
+    bool same_oneof = false;
+    if (use_custom_names && details.is_custom && match.is_custom
```

Comment:
I think either the is_custom check should be turned into an `or`.  It's very possible than the newer field has a custom json name to match the old, but they aren't both custom

Reply:
```suggestion
    if (use_custom_names && (details.is_custom || match.is_custom)
```